src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs

```diff
@@ -921,5 +921,468 @@ class C1
+                // (6,19): warning CS0219: The variable '_' is assigned but its value is never used
+                //         const int _ = 3;
+                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "_").WithArguments("_").WithLocation(6, 19)
+                );
+            CompileAndVerify(compilation, expectedOutput: "012345");
+        }
+
+        [Fact]
+        public void DiscardVsConstantInCase_04()
+        {
+            var source = @"using System;
+class Program
+{
+    static void Main()
+    {
+        const int _ = 3;
+        for (int i = 0; i < 6; i++)
+        {
+            switch (i)
+            {
+                case var _ when true:
+                    Console.Write(i);
+                    break;
+            }
+        }
+    }
+}";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics(
+                // (6,19): warning CS0219: The variable '_' is assigned but its value is never used
+                //         const int _ = 3;
+                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "_").WithArguments("_").WithLocation(6, 19)
+                );
+            CompileAndVerify(compilation, expectedOutput: "012345");
+        }
+
+        [Fact]
+        public void DiscardVsConstantInCase_05()
+        {
+            var source = @"using System;
+class Program
+{
+    static void Main()
+    {
+        const int _ = 3;
+        for (int i = 0; i < 6; i++)
+        {
+            switch (i)
+            {
+                case @_:
+                    Console.Write(i);
+                    break;
+            }
+        }
+    }
+}";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics(
+                );
+            CompileAndVerify(compilation, expectedOutput: "3");
+        }
+
+        [Fact]
+        public void DiscardVsConstantInCase_06()
+        {
+            var source = @"using System;
+class Program
+{
+    static void Main()
+    {
+        const int _ = 3;
+        for (int i = 0; i < 6; i++)
+        {
+            switch (i)
+            {
+                case @_ when true:
+                    Console.Write(i);
+                    break;
+            }
+        }
+    }
+}";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics(
+                );
+            CompileAndVerify(compilation, expectedOutput: "3");
+        }
+
+        [Fact]
+        public void DiscardVsTypeInCase_01()
+        {
+            var source = @"
+class Program
+{
+    static void Main()
+    {
+        object o = new _();
+        switch (o)
+        {
+            case _ x: break;
+        }
+    }
+}
+class _
+{
+}";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics(
+                // (9,18): error CS0119: '_' is a type, which is not valid in the given context
```

Comment:
>'_' is a type, which is not valid in the given context [](start = 41, length = 54)

This error message is a little confusing since it seems to imply that a type is not valid in this context when it is really `_` as a type that is not valid here.

Reply:
Agreed. This is an unfortunate cascaded diagnostic, but I'm not going to invest in addressing it (until someone reports finding it confusing in practice) as I don't believe it is going to realistically confuse users.

---
In reply to: [230173037](https://github.com/dotnet/roslyn/pull/30847#discussion_r230173037) [](ancestors = 230173037)